File: RetrofitReservation/app/src/main/java/ma/ensa/retrofitreservation/network/RetrofitInstance.java
package ma.ensa.retrofitreservation.network;

import com.google.gson.Gson;
import java.util.concurrent.TimeUnit;
import ma.ensa.retrofitreservation.service.ApiInterface;
import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class RetrofitInstance {

    private static Retrofit retrofit = null;
    private static final String BASE_URL = "http://192.168.0.217:8082/";

    // Configuration of the HTTP client with custom timeouts
    private static final OkHttpClient client = new OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(30, TimeUnit.SECONDS)
            .build();

    // Gson instance for custom usage
    private static final Gson gson = new Gson();

    // Method to get a Retrofit instance
    public static Retrofit getRetrofitInstance() {
        if (retrofit == null) {
            retrofit = new Retrofit.Builder()
                    .baseUrl(BASE_URL)
                    .client(client) // Adding the OkHttp client with configured timeouts
                    .addConverterFactory(GsonConverterFactory.create(gson)) // Pass custom Gson instance
                    .build();
        }
        return retrofit;
    }

    // Method to get an instance of the API Interface
    public static ApiInterface getApi() {
        return getRetrofitInstance().create(ApiInterface.class);
    }

    // Method to get the Gson instance used in Retrofit
    public static Gson getGson() {
        return gson;
    }
}

File: ReservationRetrofit_KOTLIN/app/src/main/java/ma/ensaj/reservationretrofit_kotlin/network/RetrofitInstance.kt
package ma.ensaj.reservationretrofit_kotlin.network

import ma.ensaj.reservationretrofit_kotlin.service.ApiInterface
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.util.concurrent.TimeUnit
import com.google.gson.Gson

object RetrofitInstance {

    private var retrofit: Retrofit? = null
    private const val BASE_URL = "http://192.168.0.217:8082/"

    // Configuration du client HTTP avec des délais personnalisés
    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build()

    // Gson instance for custom usage
    private val gson: Gson = Gson()

    // Méthode pour obtenir une instance de Retrofit
    @JvmStatic
    fun getRetrofitInstance(): Retrofit {
        if (retrofit == null) {
            retrofit = Retrofit.Builder()
                .baseUrl(BASE_URL)
                .client(client) // Ajout du client OkHttp avec les délais configurés
                .addConverterFactory(GsonConverterFactory.create(gson)) // Pass custom Gson instance
                .build()
        }
        return retrofit!!
    }

    // Méthode pour obtenir une instance de l'API Interface
    @JvmStatic
    fun getApi(): ApiInterface {
        return getRetrofitInstance().create(ApiInterface::class.java)
    }

    // Méthode pour obtenir l'instance Gson utilisée dans Retrofit
    @JvmStatic
    fun getGson(): Gson {
        return gson
    }
}
